File: static/FlappyClone/src/common/validation.js
// NB: If these constants are updated, remember to update the Django versions (in api/validation.py)!
const USERNAME_MIN_LENGTH = 1;
const USERNAME_MAX_LENGTH = 16;
const USERNAME_REGEX = /^[a-zA-Z0-9-_]*$/g;
const PASSWORD_MIN_LENGTH = 6;
const PASSWORD_MAX_LENGTH = 24;
const EMAIL_MIN_LENGTH = 4;
const EMAIL_MAX_LENGTH = 24;

// NB: If updating these functions, ensure that the Django functions are also updated (in api/validation.py)!
function isValidName(name) {
	var valid = true;
	var reason = "";
	var help = "";
	if (typeof name !== "string") {
		valid = false;
		reason = "Please enter a username.";
	} else if (name === "") {
		valid = false;
		reason = "Please enter a username.";
	} else if (name.length < USERNAME_MIN_LENGTH) {
		valid = false;
		reason = "Username is too short.";
	} else if (name.length > USERNAME_MAX_LENGTH) {
		valid = false;
		reason = "Username is too long.";
	} else if ((name.match(USERNAME_REGEX) || []).length === 0) {
		valid = false;
		reason = "Invalid username.";
		help = "Username must contain only alphanumeric characters, dashes or underscores.";
	}
	return {valid: valid, reason: reason, help: help};
}

// NB: If updating these functions, ensure that the Django functions are also updated (in api/validation.py)!
function isValidPassword(password) {
	var valid = true;
	var reason = "";
	var help = "";
	if (typeof password !== "string") {
		valid = false;
		reason = "Please enter a password.";
	} else if (password === "") {
		valid = false;
		reason = "Please enter a password.";
	} else if (password.length < PASSWORD_MIN_LENGTH) {
		valid = false;
		reason = "Password is too short.";
	} else if (password.length > PASSWORD_MAX_LENGTH) {
		valid = false;
		reason = "Password is too long.";
	}
	return {valid: valid, reason: reason, help: help};
}

// NB: If updating these functions, ensure that the Django functions are also updated (in api/validation.py)!
function isValidEmail(email) {
	var valid = true;
	var reason = "";
	var help = "";
	if (typeof email !== "string") {
		valid = false;
		reason = "Please enter an email.";
	} else if (email === "") {
		valid = false;
		reason = "Please enter an email.";
	} else if (email.length < EMAIL_MIN_LENGTH) {
		valid = false;
		reason = "Email is too short.";
	} else if (email.length > EMAIL_MAX_LENGTH) {
		valid = false;
		reason = "Email is too long.";
	} else {
		var num = (email.match(/@/g) || []).length;
		if (num !== 1) {
			valid = false;
			reason = "Invalid email.";
		}
	}
	return {valid: valid, reason: reason, help: help};
}

File: api/validation.py
from django.core.validators import MinLengthValidator, MaxLengthValidator, RegexValidator, EmailValidator

# NB: If these constants are updated, remember to update the JavaScript versions (in js/game/validation.js)!!
NUM_LEADERBOARD_ENTRIES = 10
MAX_NAME_LENGTH = 16
LEGAL_SYMBOLS = '-_'

# NB: If updating these functions, ensure that the JavaScript functions are also updated (in js/game/validation.js)!
def is_valid_name_char(c):
    if (c >= 'a' and c <= 'z'):
        return True
    elif (c >= 'A' and c <= 'Z'):
        return True
    elif (c >= '1' and c <= '9'):
        return True
    elif (LEGAL_SYMBOLS.find(c) != -1):
        return True
    else:
        return False

# NB: If updating these functions, ensure that the JavaScript functions are also updated (in js/game/validation.js)!
def is_valid_name(name):
    reason = ''
    if (len(name) == 0):
        return (False, 'name is an empty string')
    elif (len(name) > MAX_NAME_LENGTH):
        return (False, 'name is too long (' + len(name) + ' characters, max is ' + MAX_NAME_LENGTH + ')')
    else:
        for c in name:
            if (not is_valid_name_char(c)):
                return (False, 'name contains an invalid character (' + c + ')')
    return True
